### PR TITLE
Test untested functions and traits of  `Felt`

### PR DESCRIPTION
**Pull request by haroune-mohammedi**
## Description
Issue: #823

We added three tests for `num_traits::One`:

- [x] `one_value` to test `One::one`
- [x] `is_one` to test `One::is_zero`
- [x] `one_multiplicative_identity` to test the multiplicative identity:

```
x * 1 = x       ∀ x
1 * x = x       ∀ x
```
- [x] `zero_additive_identity` to test the additive identity: of `num_traits::Zero`:
``` 
x + 0 = x       ∀ x
0 + x = x       ∀ x
```

Tests for `num_traits::Signed` implementation
   
- [x] `felt_is_never_negative`
- [x] `non_zero_felt_is_always_positive`
- [x]  `non_zero_felt_signum_is_always_one`
- [x] `signum_of_zero_is_zero`
- [x]  `sub_abs`
- [x]  `abs`

Tests for `Integer`:
- [x] `divides_doesnt_panic`
- [x] `gcd_doesnt_panic`
- [x] `is_even_always_true`
- [x] `is_odd_always_false`

Tests for `FromPrimitive` and `ToPrimitive`:
- [x] `from_u64_and_to_u64_primitive`
- [x] `from_i64_and_to_i64_primitive`

Tests for:
- [x] `modpow_in_range`
- [x] `sqrt_in_range`
- [x] `sqrt_is_inv_square`

Tests for `Add`,  `AddAssign`,  `Add<u32>`, `Sub<u32>`, `Sub<usize>`:

- [x] `add_u32_in_range`
- [x] `add_u32_is_inv_sub`
- [x] `sub_u32_in_range`
- [x] `sub_u32_is_inv_add`
- [x] `sub_uszie_in_range`
- [x] `sub_usize_is_inv_add`
- [x] `add_in_range`
- [x] `add_is_inv_sub`
- [x] `add_assign_in_range`

We tested the reference variant for 

- [x] `sub_assign_in_range`
- [x] `neg_in_range`
- [x] `pow_in_range`

As a result of writing tests, we discoverd bugs and fixed the implementation of `is_even`, `is_odd` and `divides`

## Checklist
- [x] Linked to Github Issue
- [x] Unit tests added
- [ ] Integration tests added.
- [ ] This change requires new documentation.
  - [ ] Documentation has been added/updated.

### CODE REVIEW

**Review by fmoletta — APPROVED** [review]
LGTM

**Review by Jrigada — APPROVED** [review]
Thanks for the PR, LGTM

**Comment by haroune-mohammedi** [comment]
I have some suggestions that involves changes to the tests I haven't wrote so I figured I should ask you guys first

- Use `Felt::max_value` instead of parsing `PRIME_STR` inside all the functions
- use ref implementation to implement owned ones instead of implementing the same thing twice
- Why are we only testing for in_range, why don't we test the result values of the operations like: add, neg, sub ..., 
- Why do we need to implement the same operation for `Felt`, `&Felt`, `u32`, `usize` ... check https://docs.rs/impl_ops/0.1.1/impl_ops/index.html that could be used to get rid of the boilerplate

**Comment by haroune-mohammedi** [comment]
I've added tests for `Felt` and updated the description

@fmoletta @Jrigada would you please re-review and approve the workflows, also is there any way I can run the workflows without waiting approval in my fork ?

File: felt/src/lib.rs
Comment (Oppen) [inline felt/src/lib.rs:1091]: Let's remove this.

File: felt/src/lib.rs
Comment (Oppen) [inline felt/src/lib.rs:1098]: This test will fail for `y == 0`. I'm not 100% what should happen for `x == 0` if this is not returning a `Result<_>`, but we're passing it in the test as well.

File: felt/src/lib.rs
Comment (Oppen) [inline felt/src/lib.rs:1113]: This will fail for a bunch of even uints. Take `CAIRO_PRIME + 1` for example. That's even before taking the modulus, but it becomes `1` afterwards.

File: felt/src/lib.rs
Comment (Oppen) [inline felt/src/lib.rs:1119]: The same happens here, `CAIRO_PRIME + 2` is odd, but `CAIRO_PRIME + 2 mod CAIRO_PRIME` is `2` which is even.

File: felt/src/lib.rs
Comment (Oppen) [inline felt/src/lib.rs:1148]: Nit: add a newline.

File: felt/src/lib.rs
Comment (Oppen) [inline felt/src/lib.rs:1167]: IIRC, the `is_positive`, `signum` and `is_negative` for type `Felt` are meant for the signed representation, which reads a field element as the range `(-CAIRO_PRIME/2, CAIRO_PRIME/2)`, where `x > CAIRO_PRIME/2` represents a negative field element. This would make these assumptions wrong.

Comment (Oppen) [reply felt/src/lib.rs:1167]: There was the intent to make a separate type for singed field elements to avoid this confusion, I don't exactly remember if it was applied (I'll check later today and come back with a definitive answer).

File: felt/src/lib.rs
Comment (Oppen) [inline felt/src/lib.rs:1183]: For the same reason as above, I'm not sure if this is true.

File: felt/src/lib.rs
Comment (Oppen) [inline felt/src/lib.rs:1223]: `expected_y` does not necessarily fit in a `uint32`. `Felt(CAIRO_PRIME - 1) - u32::MAX` is still greater than `u32::MAX`.

Comment (Oppen) [reply felt/src/lib.rs:1223]: You should convert `y` to `BigUint` before the comparison instead.

File: felt/src/lib.rs
Comment (Oppen) [inline felt/src/lib.rs:1230]: Don't we already have this prime parsed?

File: felt/src/lib.rs
Comment (Oppen) [inline felt/src/lib.rs:1244]: Ditto.

File: felt/src/lib.rs
Comment (Oppen) [inline felt/src/lib.rs:1259]: Ditto.

File: felt/src/lib.rs
Comment (Oppen) [inline felt/src/lib.rs:1279]: Ditto.

**Review by Oppen — CHANGES_REQUESTED** [review]
Hello @haroune-mohammedi, thanks for your contribution, it looks good.
However, there are a few changes I need you to make before we can merge it.

File: felt/src/lib.rs
Comment (Oppen) [reply felt/src/lib.rs:1167]: Effectively it still works as I detailed in the first comment.

File: felt/src/lib.rs
Comment (haroune-mohammedi) [reply felt/src/lib.rs:1098]: It won't fail, `is_multiple_of` and `divides` always return `true`, see [`bigint_felt.rs#L615`](https://github.com/Quadratic-Labs/cairo-rs/blob/ef74e80fe4f8f1156233418e3a2c381db5b14047/felt/src/bigint_felt.rs#L615)

File: felt/src/lib.rs
Comment (haroune-mohammedi) [reply felt/src/lib.rs:1119]: Same answer as `is_even`

File: felt/src/lib.rs
Comment (haroune-mohammedi) [reply felt/src/lib.rs:1167]: what should be done about them ?

File: felt/src/lib.rs
Comment (haroune-mohammedi) [reply felt/src/lib.rs:1223]: Why wouldn't it fit, we're subtracting `x` from `x`, we should be left with exactly `y` or am I missing something here ?

File: felt/src/lib.rs
Comment (haroune-mohammedi) [reply felt/src/lib.rs:1230]: it's being parsed in every function, I wanted to get it out or even use `Felt::max_value` instead but I didn't want to change before asking (I asked in the comments about it)

File: felt/src/lib.rs
Comment (Oppen) [reply felt/src/lib.rs:1113]: Sorry, I thought you were comparing with the `BigUint` created from the string, not from the `Felt` itself.

File: felt/src/lib.rs
Comment (Oppen) [reply felt/src/lib.rs:1223]: Ugh, didn't see the `- x` 🤦

File: felt/src/lib.rs
Comment (Oppen) [reply felt/src/lib.rs:1098]: That is odd. `0` shouldn't divide anything.

File: felt/src/lib.rs
Comment (Oppen) [reply felt/src/lib.rs:1167]: I'd say the practical thing is to split each test in two and use a `prop_assume!` to fix the range. One test checks for negatives, one test checks for positives. The `0` case can be made a normal unit test.

Comment (haroune-mohammedi) [reply felt/src/lib.rs:1167]: But that's not what is implemented in `Signed`, do you mean we have to change the implementation and make felts where `x > CAIRO_PRIME/2` negative ? What about having a separate `SignedFelt` type, isn't it a better design than messing with the implementation of `Felt` ?

Comment (Oppen) [reply felt/src/lib.rs]: Yes, the idea was to eventually refactor into a `SignedFelt` type that you could build from `Felt`. But for now it is that way. It should be documented and tested as that until we do the refactor. We're aware as it is right now it's unintuitive, but we prioritized fixing quickly fast over having it be beautiful, with our sight on refactoring it afterwards.
If you'd rather have it that way __now__ so tests can be more obvious, we can either see if someone on our side has time to do the refactor now (I think everyone has a task currently, so this is possible but unlikely), or maybe we can create an issue and you can assign it to yourself if you wish. The task would be to track where we actually care about sign (it's a few hints IIRC) as well as create the base shim.

**Review by Oppen — APPROVED** [review]

**Pull request merged**
